### PR TITLE
[Black Jack]: Comparisons Concept and Concept Exercise Finalization

### PR DESCRIPTION
**Pull request by BethanyG**
Added/Changed/Edited files:

- [x]   config.json
- [x]   exercises/concept/black-jack/.meta/config.json
----
- [x]   concepts/comparisons/introduction.md
- [x]   concepts/comparisons/about.md
____
- [x]   exercises/concept/black-jack/.docs/introduction.md
- [x]   exercises/concept/black-jack/.docs/instructions.md
- [x]   exercises/concept/black-jack/.docs/hints.md
____
- [x]  exercises/concept/black-jack/.meta/exemplar.py
- [x]  exercises/concept/black-jack/black_jack.py
- [x]   exercises/concept/black-jack/black_jack_test.py

### CODE REVIEW

File: concepts/comparisons/about.md
Comment (kotp) [inline concepts/comparisons/about.md:17]: If we are alligning for the table we should do so.

File: concepts/comparisons/about.md
Comment (kotp) [inline concepts/comparisons/about.md:188]: If the `<>` for this is meant to be code, we should do something similar as done in the "and only if" statement where it is wrapped in code delimiters inline.

File: concepts/comparisons/about.md
Comment (kotp) [inline concepts/comparisons/about.md:189]: It looks like the code delimiter changed here as well, perhaps unintended.

File: concepts/comparisons/introduction.md
Comment (kotp) [inline concepts/comparisons/introduction.md:17]: Same here for alignment of column delimiter.

File: concepts/comparisons/introduction.md
Comment (kotp) [inline concepts/comparisons/introduction.md:24]: Two blank spaces which is usually ignored in HTML rendering, and not really helpful in the plain text viewing.  This happens in other places as well.  Single blank line between parenthesis is likely intended.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (kotp) [inline exercises/concept/black-jack/.docs/introduction.md:16]: Another column delimiter off kilter.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (kotp) [inline exercises/concept/black-jack/.docs/introduction.md:157]: Starting to feel like this is coming from some place higher up the chain maybe, as this seems like it should be

    apple is orange

We can compare apples to oranges, and those would be valid names, but not sure what the `<>` does to them, though I think the communication that I get from reading about apples and oranges are a generation, kind of like `foo` or `bar`, and maybe the `<foo>` and `<bar>` does something different in Python specifically.

So maybe take the feedback like you might with apple slices, you know, with a little salt.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (kotp) [inline exercises/concept/black-jack/.docs/introduction.md:158]: I think this should be 

```markdown
`apple is not orange` yields the inverse.
```

instead.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (kotp) [inline exercises/concept/black-jack/.docs/introduction.md:160]: Maybe expand on why, other than only "because of their singleton status".  I am sure there is something more implied here.  And I can not think of why I would ever do other comparisons, anyway.  So the fact that it is mentioned and specifically mentioned that they are singletons, raises an eyebrow, wanting more information.

File: exercises/concept/black-jack/.meta/exemplar.py
Comment (kotp) [inline exercises/concept/black-jack/.meta/exemplar.py:6]: Why a blank line around this string, and not the following?

File: exercises/concept/black-jack/black_jack.py
Comment (kotp) [inline exercises/concept/black-jack/black_jack.py:6]: Also blank line around the string content unlike some other `"""` strings.

**Review by kotp — CHANGES_REQUESTED** [review]
Looking pretty good, but I think there may have been some unintended changes, plus what appears to be style deviance instances in the documents themselves.

And a question about `<apples>` and `<oranges>`.

File: concepts/comparisons/about.md
Comment (BethanyG) [reply concepts/comparisons/about.md:17]: There's a big difference in the way this appears in code vs render.  When rendered, this is aligned: 

| Operator | Operation                | Description                                                    |
| -------- | ------------------------ | -------------------------------------------------------------- |
| `>`      | "greater than"             | `a > b` is `True` if `a` is **strictly** greater in value than `b` |
| `<`      | "less than"                | `a < b` is `True` if `a` is **strictly** less in value than `b`    |
| `==`     | "equal to"                 | `a == b` is `True` if `a` is **strictly** equal to `b` in value    |
| `>=`     | "greater than or equal to" | `a >= b` is `True` if `a > b` OR `a == b` in value             |
| `<=`     | "less than or equal to"    | `a <= b` is `True` if `a < b` or `a == b` in value             |
| `!=`     | "not equal to"             | `a != b` is `True` if `a == b` is `False`                      |
| `is`     | "identity"                 | `a is b` is `True` if **_and only if_** `a` and `b` are the same _object_  |
| `is not` | "negated identity"                 | `a is not b` is `True` if `a` and `b` are **not** the same _object_ |
| `in`     | "containment test"         | `a in b` is `True` if `a` is member, subset, or element of `b`               |
| `not in` | "negated containment test" | `a not in b` is `True` if `a` is not a member, subset, or element of `b`     |

File: concepts/comparisons/about.md
Comment (BethanyG) [reply concepts/comparisons/about.md:188]: Yeah. I debated that.  This way, it emphasizes the `is` operator, but requires I escape the `<`.  Not sure I like how `<apple> is <orange>` reads as one block.  We _could_ do `<apple>` `is` `<orange>` ...dunno if that is better.  Thoughts?

File: concepts/comparisons/about.md
Comment (BethanyG) [reply concepts/comparisons/about.md:189]: Same as above -- not sure what reads better, but a whole block doesn't read well to me.

File: concepts/comparisons/introduction.md
Comment (BethanyG) [reply concepts/comparisons/introduction.md:17]: Same comment as above.  The render is more important than the code -- and the render is in alignment.

File: concepts/comparisons/introduction.md
Comment (BethanyG) [reply concepts/comparisons/introduction.md:24]: I am not sure I follow here.  Which two spaces are we talking about?

File: exercises/concept/black-jack/.docs/introduction.md
Comment (BethanyG) [reply exercises/concept/black-jack/.docs/introduction.md:157]: 😆   So..same comment here as I had earlier.  We just need to decide which reads better.  The _intent_ of the `<` is to denote something similar to `<foo>` and `<bar>` - placeholders -- just not `foo` and `bar` placeholders.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (BethanyG) [reply exercises/concept/black-jack/.docs/introduction.md:158]: yeah..or whatever we end up deciding.  Although it should be `<apple> is not <orange>` -- if we want to keep to the `<>` signifying a variable or object.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (BethanyG) [reply exercises/concept/black-jack/.docs/introduction.md:160]: I can put a link into the python docs.  We're practically quoting them here.  The `singleton status` -- is exactly that -- there is only ever ONE instance of either `None` or `NotImplemented` in any running Python program.  So equivalence is never the right metaphor.....it is either **THE** `None` or **THE** `NotImplemented` - nothing is ever equal to them, because nothing can ever BE them .... except themselves.

Comment (BethanyG) [reply exercises/concept/black-jack/.docs/introduction.md:160]: Here is the relevant snip:

Comment (BethanyG) [reply exercises/concept/black-jack/.docs/introduction.md:160]: ..I guess we could also link to PEP8 - but the length of that document might confuse someone who is trying to just learn about comparisons.

File: exercises/concept/black-jack/.meta/exemplar.py
Comment (BethanyG) [reply exercises/concept/black-jack/.meta/exemplar.py:6]: [PEP257](https://www.python.org/dev/peps/pep-0257/) puts the closing quotes on their own lines.  PyCharm defaults to inserting and extra line above - and since I think it also reads better that way, I left it.

File: concepts/comparisons/about.md
Comment (BethanyG) [reply concepts/comparisons/about.md:17]: Neverimind.  Prettier to the rescue!

Comment (kotp) [reply concepts/comparisons/about.md:17]: Yep, I knew that it would look good rendered (markdown is awesome, tables are cool that way), but having it look good in source is nice too.

The other choice is "less white space" and not align any of it, but have it naturally end, which I do not _not_ like choice as well.

But having it inconsistent, well, that is distracting. ;)

File: exercises/concept/black-jack/.meta/exemplar.py
Comment (BethanyG) [reply exercises/concept/black-jack/.meta/exemplar.py:6]: I'll normalize it.

**Comment by BethanyG** [comment]
Alrighty.  Ran everything through priettier, and cleaned up the `<apples>` and `<oranges>` 😉   I also removed the excess lines in the module docstrings.  Let me know if there's anything I've missed.  😄   Thanks for reviewing this!

File: exercises/concept/black-jack/.docs/introduction.md
Comment (kotp) [reply exercises/concept/black-jack/.docs/introduction.md:149]: I like that, I think the link to that section is good.  I like the addition of "never the equality operators".  A little more prodding to the right idea here.  That we check for identity not equality when asking if `ourThing is None`, or if `ourBehavior is not NotI'mplemented` if that is correct.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (kotp) [reply exercises/concept/black-jack/.docs/introduction.md]: If that `<>` indicates variable or object in Python speak then that is what should be used.  I think in the context, perhaps it is clear to focus on the `is not` rather than the variable name used (and is why we use foo and bar as well known "can be whatever, as allowed by the syntax of the language in that case sensitive interpretation".

File: concepts/comparisons/introduction.md
Comment (kotp) [reply concepts/comparisons/introduction.md]: Two blank vertical spaces (new lines) is what I really mean here.

File: concepts/comparisons/about.md
Comment (kotp) [reply concepts/comparisons/about.md]: I think that I notice the delimiters  and generally ignore the name of the variables being used anyway.  So to me the `<>` is distracting from the communication, while if I saw `apple is orange` where the reading is talking about `is` and `is not` I would see the example as intended, that the names on either side do not really matter, the "how to use the identity operator `is` or `is not`.  It is why we see `foo` and `bar` as nonsense names in code documentation, we learn that those are not the important part, only that something is there in each place.

File: exercises/concept/black-jack/.docs/introduction.md
Comment (kotp) [reply exercises/concept/black-jack/.docs/introduction.md]: Right, but in this case, I see the `<>` and that yells pay attention to this, because that is what delimiters do for me in code.

**Comment by BethanyG** [comment]
@kotp - are there additional changes needed here for approval?  Let me know - thanks!

**Review by kotp — APPROVED** [review]

**Comment by kotp** [comment]
I approve, as long as `<fish>` is valid Python.  I kind of think it may be confusing if the student plus that in and it does not work.  And then has to interpret it more as Unix documentation style for the meanings of those brackets.

**Comment by BethanyG** [comment]
I have no bead on if `<fish>` is going to be confusing.  The convention is used in many Python concept exercises to denote a placeholder variable or parameter, to signal to the student that the _name_ can really be anything they want.

My thought is to have the exercise go live in `beta` mode, and if it is confusing to students they will (loudly) let us know - and we can discuss re-working it at that point.

**Comment by kotp** [comment]
On 22/01/08 10:27AM, BethanyG wrote:
> I have no bead on if `<fish>` is going to be confusing.  The
> convention is used in many Python concept exercises to denote a
> placeholder variable or parameter, to signal to the student that the
> _name_ can really be anything they want.

Also, if it has been being used, and understood well,
already, then I do not have a problem with it.

> My thought is to have the exercise go live in `beta` mode, and if it
> is confusing to students they will (loudly) let us know - and we can
> discuss re-working it at that point.

Definitely.  I do not know if this is common outside of the Python
track, and used widely in the Python greater community.  As an
"outsider" I definitely that as a perspective.  But if it is convention
in the Python community, then it should be used here as well, as the
student will learn that this is not something unique in how this is
communicated on Exercism.

**Pull request merged**
